src/docfx/lib/ProcessUtility.cs

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Docs
+{
+    internal static class ProcessUtility
+    {
+        /// <summary>
+        /// Execute process with args
+        /// </summary>
+        /// <param name="fileName">The process path name or location</param>
+        /// <param name="commandLineArgs">The process command line args</param>
+        /// <param name="cwd">The current working directory</param>
+        /// <param name="timeout">The timeout setting</param>
+        /// <returns>The executed result</returns>
+        public static Task<string> Execute(string fileName, string commandLineArgs, string cwd = null, TimeSpan? timeout = null)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(fileName));
+
+            var tcs = new TaskCompletionSource<string>();
+
+            var error = new StringBuilder();
+            var output = new StringBuilder();
+
+            var process = new Process
+            {
+                EnableRaisingEvents = true,
+                StartInfo = new ProcessStartInfo
+                {
+                    FileName = fileName,
+                    WorkingDirectory = cwd,
+                    Arguments = commandLineArgs,
+                    CreateNoWindow = true,
+                    UseShellExecute = false,
+                    RedirectStandardOutput = true,
+                    RedirectStandardError = true,
+                },
+            };
+
+            process.OutputDataReceived += (sender, e) => output.AppendLine(e.Data);
+            process.ErrorDataReceived += (sender, e) => error.AppendLine(e.Data);
+
+            process.Exited += (a, b) =>
+            {
+                // Wait for exit here to ensure the standard output/error is flushed.
+                process.WaitForExit();
+
+                if (process.ExitCode == 0)
+                {
+                    tcs.TrySetResult(output.ToString().Trim());
+                }
+                else
+                {
+                    var message = $"'\"{fileName}\" {commandLineArgs}' failed in directory '{cwd}' with exit code {process.ExitCode}: \nSTDOUT:'{output}'\nSTDERR:\n'{error}'";
+
+                    tcs.TrySetException(new InvalidOperationException(message));
+                }
+            };
+
+            Task.Delay(timeout ?? TimeSpan.FromMinutes(5)).ContinueWith(
+                task =>
+                {
+                    if (!process.HasExited)
+                    {
+                        process.Kill();
+                    }
+
+                    tcs.TrySetCanceled();
```

Comment:
>TrySetCanceled [](start = 24, length = 14)

shall we throw TimeOut exception?

Reply:
i believe they represent the same purpose

---
In reply to: [180617087](https://github.com/dotnet/docfx/pull/2611#discussion_r180617087) [](ancestors = 180617087)

Reply:
Can we have more user-friendly error message with TaskCancelledException?

---
In reply to: [180620087](https://github.com/dotnet/docfx/pull/2611#discussion_r180620087) [](ancestors = 180620087,180617087)

Reply:
OK

---
In reply to: [180623688](https://github.com/dotnet/docfx/pull/2611#discussion_r180623688) [](ancestors = 180623688,180620087,180617087)